File: src/akinator/Akinator.java
package akinator;

import java.util.Scanner;

public class Akinator {

    private static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        String cadena="$cn_3i3gramm7reenait$",nuevaCadena="";
        char array[]=cadena.toCharArray();
        for(char Char:array){
            nuevaCadena+=(((int)Char)+5)+" ";
        }
        
        String ArrayString[]=nuevaCadena.split(" "),nuevaCadena1="";
        for(String string:ArrayString){
            int entero=Integer.parseInt(string);
            nuevaCadena1+=(char)(entero);
        }
        System.out.println("Ascii: "+nuevaCadena);
        System.out.println("Cadena: "+nuevaCadena1);
        
    
        Akinator ak = new Akinator();
        boolean newGame = true;
        while (newGame) {
            ak.welcome();
            String start = sc.next();
            if (start.equals("no") || start.equals("n")) {
                newGame = false;
                continue;
            }

            Sql db = new Sql();
            String[][] datosHeroes = db.getMatriz();

            int sizeHeroes = datosHeroes.length;
            int sizeArbol = 0;

            Nodo nodo = new Nodo(datosHeroes[0][0], datosHeroes[0][1]);
            Nodo root = nodo;
            boolean victoria = false;
            while (sizeArbol < sizeHeroes && (!victoria)) {
                sizeArbol++;
                System.out.print("\n¿Tu superheroe " + nodo.getCualidad() + "?\n> ");
                String respuesta = sc.next();

                if (respuesta.equals("si") || respuesta.equals("s")) {
                    System.out.print("\n¿Tu superheroe es " + nodo.getNombre() + "?\n> ");
                    respuesta = sc.next();
                    if (respuesta.equals("si") || respuesta.equals("s")) {
                        nodo.setNodoIzquierdo(nodo);
                        ak.Victoria();
                        victoria = true;
                    } else {
                        if (sizeArbol == sizeHeroes) {
                            String[] datos = ak.Derrota();
                            nodo.setNodoIzquierdo(new Nodo(datos[0], datos[1]));
                            nodo = nodo.getNodoIzquierdo();
                            db.insert(datos);
                            continue;
                        }

                        nodo.setNodoDerecho(new Nodo(datosHeroes[sizeArbol][0], datosHeroes[sizeArbol][1]));
                        nodo = nodo.getNodoDerecho();
                    }
                } else {
                    if (sizeArbol == sizeHeroes) {
                        String[] datos = ak.Derrota();
                        db.insert(datos);
                        continue;
                    }

                    nodo.setNodoDerecho(new Nodo(datosHeroes[sizeArbol][0], datosHeroes[sizeArbol][1]));
                    nodo = nodo.getNodoDerecho();
                }
            }
            //ak.printRecorrido(root, sizeArbol, victoria);
            String desicion = ak.StartAgain();
            if (desicion.equals("no") || desicion.equals("n")) {
                newGame = false;
            }
        }
        System.out.println("\nGracias por jugar\n");
    }

    public void welcome() {
        System.out.println("-----------------------------");
        System.out.println("|Instrucciones              |");
        System.out.println("|Verdadero = \'si\' o \'s\'     |");
        System.out.println("|Falso = \'no\' o \'n\'         |");
        System.out.println("-----------------------------\n");
        System.out.print("Piensa en un superheroe\ny yo tratare de adivinar cual es, ¿Estas listo?\n> ");
    }

    public void Victoria() {
        System.out.println("\n-------------");
        System.out.println("|VICTORIA!!!|");
        System.out.println("-------------");
    }

    public String[] Derrota() {
        System.out.println("\nMe rindo");
        System.out.print("\n¿Dime cual es el nombre de tu superheroe?\n> ");
        sc.nextLine();
        String heroe = sc.nextLine();
        System.out.print("\n¿Dime cual es su cualidad?\n> ");
        String cualidad = sc.nextLine();
        return new String[]{heroe, cualidad};
    }

    public String StartAgain() {
        System.out.print("\n¿Quieres jugar otra vez?\n> ");
        return sc.next();
    }

    public void printRecorrido(Nodo nodo, int size, boolean estado) {
        for (int i = 0; i < size; i++) {
            System.out.println(nodo.getNombre());
            if (nodo.getNodoIzquierdo() != null) {
                nodo = nodo.getNodoIzquierdo();
                System.out.println("  /");
            } else {
                nodo = nodo.getNodoDerecho();
                System.out.println("  \\");
            }
        }
        if (estado) {
            System.out.println("victoria");
        } else {
            System.out.println("derrota");
        }
    }
}

File: pyhton/Akinator/akinator.py
from Nodo import Nodo
from Sql import database as DB

def welcome():
    print("-----------------------------")
    print("|Instrucciones              |")
    print("|Verdadero = \'si\' o \'s\'     |")
    print("|Falso = \'no\' o \'n\'         |")
    print("-----------------------------\n")
    print("Piensa en un superheroe\ny yo tratare de adivinar cual es, ¿Estas listo?")

def Victoria():
    print("\n-------------")
    print("|VICTORIA!!!|")
    print("-------------")

def Derrota():
    print("\nMe rindo")
    print("\n¿Dime cual es el nombre de tu superheroe?")
    heroe = input("> ")
    print("\n¿Dime cual es su cualidad?")
    cualidad = input("> ")
    return heroe,cualidad

def StartAgain():
    print("\n¿Quieres jugar otra vez?")
    return input("> ")

def printRecorrido(nodo,size,estado):
    for i in range(size):
        print(nodo.getNombre())
        if nodo.getNodoIzquierdo() != None:
            nodo = nodo.getNodoIzquierdo()
            print("  /")
        else:
            nodo = nodo.getNodoDerecho()
            print("  \\")
    if estado: #if estado = victoria
        print("victoria")
    else:
        print("derrota")


if __name__ == "__main__":
    newGame = True
    while newGame:
        welcome()
        start = input("> ")
        if(start == 'no' or start == 'n'):
            newGame = False
            continue
        
        db = DB()
        datosHeroes = db.getMatriz()

        sizeHeroes = len(datosHeroes)
        sizeArbol = 0

        nodo = Nodo(datosHeroes[0][0],datosHeroes[0][1])
        root = nodo
        victoria = False
        while sizeArbol < sizeHeroes and not victoria:
            sizeArbol += 1
            print(f"\n¿Tu superheroe {nodo.getCualidad()}?")
            respuesta = input("> ")

            if respuesta == 'si' or respuesta == 's':
                print(f"\n¿Tu superheroe es {nodo.getNombre()}?")
                respuesta = input("> ")
                if respuesta == 'si' or respuesta == 's':
                    nodo.setNodoIzquierdo(nodo)
                    Victoria()
                    victoria = True
                else:
                    if sizeArbol == sizeHeroes:
                        datos = Derrota()
                        nodo.setNodoIzquierdo(Nodo(datos[0],datos[1]))
                        nodo = nodo.getNodoIzquierdo()
                        db.insert(datos)
                        continue

                    nodo.setNodoDerecho(Nodo(datosHeroes[sizeArbol][0],datosHeroes[sizeArbol][1]))
                    nodo = nodo.getNodoDerecho()
            else:
                if sizeArbol == sizeHeroes: #if we ran out of heroes
                    datos = Derrota()
                    db.insert(datos)
                    continue
                
                nodo.setNodoDerecho(Nodo(datosHeroes[sizeArbol][0],datosHeroes[sizeArbol][1]))
                nodo = nodo.getNodoDerecho()

        printRecorrido(root,sizeArbol,victoria)
        desicion = StartAgain()
        if desicion == "no" or desicion == "n":
            newGame = False
    
    print("\nGracias por jugar\n")

#C:\Users\ralfr\Desktop\Files\Python\Akinator\akinator.py